File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/forge/ForgeDownloadInfo.java
package org.wdt.wdtc.core.download.forge;


import com.google.gson.JsonObject;
import lombok.Getter;
import lombok.SneakyThrows;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.download.infterface.DownloadSourceInterface;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.ModUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;
import org.wdt.wdtc.core.utils.ZipUtils;

import java.io.File;
import java.io.IOException;

public class ForgeDownloadInfo implements ModDownloadInfoInterface {
  private static final Logger logmaker = WdtcLogger.getLogger(ForgeDownloadInfo.class);
  protected final DownloadSourceInterface source;
  @Getter
  protected final String ForgeVersionNumber;
  protected final Launcher launcher;


  public ForgeDownloadInfo(Launcher launcher, String forgeVersionNumber) {
    ForgeVersionNumber = forgeVersionNumber;
    this.launcher = launcher;
    this.source = DownloadSourceManger.getDownloadSource();
  }

  public ForgeDownloadInfo(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    this(launcher, versionJsonObjectInterface.getVersionNumber());
  }

  public void DownloadInstallJar() {
    DownloadUtils.StartDownloadTask(getForgeInstallJarUrl(), getForgeInstallJarPath());
  }

  private String getInstallJarUrl() {
    return source.getForgeLibraryMavenUrl() + "net/minecraftforge/forge/:mcversion-:forgeversion/forge-:mcversion-:forgeversion-installer.jar";
  }

  public File getForgeInstallJarPath() {
    return new File(FileManger.getWdtcCache(), ForgeVersionNumber + "-installer.jar");
  }

  public String getForgeInstallJarUrl() {
    return getInstallJarUrl().replaceAll(":mcversion", launcher.getVersionNumber()).replaceAll(":forgeversion", ForgeVersionNumber);
  }

  public void getInstallProfile() {
    ZipUtils.unZipToFile(getForgeInstallJarPath(), getInstallProfileFile(), "install_profile.json");

  }


  public File getInstallProfileFile() {
    return FileUtils.toFile(FileManger.getWdtcCache(), "/install_profile" + "-" + launcher.getVersionNumber() + "-" + ForgeVersionNumber + ".json");
  }

  public JsonObject getInstallPrefileJsonObject() throws IOException {
    return JsonUtils.readFileToJsonObject(getInstallProfileFile());
  }


  public File getForgeVersionJsonFile() {
    return new File(FileManger.getWdtcCache(), "version-" + launcher.getVersionNumber() + "-" + ForgeVersionNumber + ".json");
  }

  public void getForgeVersionJson() {
    ZipUtils.unZipToFile(getForgeInstallJarPath(), getForgeVersionJsonFile(), "version.json");
  }

  @SneakyThrows
  public JsonObject getForgeVersionJsonObject() {
    return JsonUtils.readFileToJsonObject(getForgeVersionJsonFile());
  }


  @Override
  public String getModVersion() {
    return ForgeVersionNumber;
  }

  @Override
  public InstallTaskInterface getModInstallTask() {
    return new ForgeInstallTask(launcher, ForgeVersionNumber);
  }

  @Override
  public ModUtils.KindOfMod getModKind() {
    return ModUtils.KindOfMod.FORGE;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/forge/ForgeDownloadInfo.kt
package org.wdt.wdtc.core.download.forge

import com.google.gson.annotations.JsonAdapter
import org.wdt.utils.gson.readFileToClass
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.game.serializeGameRuntimeDependencyGsonBuilder
import org.wdt.wdtc.core.manger.currentDownloadSource
import org.wdt.wdtc.core.manger.wdtcCache
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.gson.DownloadInfoTypeAdapter
import org.wdt.wdtc.core.utils.startDownloadTask
import org.wdt.wdtc.core.utils.toURL
import org.wdt.wdtc.core.utils.unZipToFile
import java.io.File
import java.net.URL

// TODO Using 'Url' class
@JsonAdapter(DownloadInfoTypeAdapter::class)
open class ForgeDownloadInfo(protected val version: Version, override val modVersion: String) :
	ModDownloadInfoInterface {
	
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) :
		this(version, versionsJsonObjectInterface.versionNumber)
	
	suspend fun startDownloadInstallJar() {
		startDownloadTask(forgeInstallJarUrl, forgeInstallJarFile)
	}
	
	val forgeInstallJarFile: File
		get() = File(wdtcCache, "$modVersion-installer.jar")
	private val forgeInstallJarUrl: URL
		get() = version.versionNumber.let {
			"${currentDownloadSource.forgeLibraryMavenUrl}net/minecraftforge/forge/${it}-$modVersion/forge-$it-$modVersion-installer.jar".toURL()
		}
	
	fun unZipToInstallProfile() {
		unZipToFile(forgeInstallJarFile, forgeInstallProfileFile, "install_profile.json")
	}
	
	val forgeInstallProfileFile: File
		get() = File(wdtcCache, "install_profile-${version.versionNumber}-$modVersion.json")
	
	val forgeVersionJsonFile: File
		get() = File(wdtcCache, "version-${version.versionNumber}-$modVersion.json")
	
	fun unZipToForgeVersionJson() {
		unZipToFile(forgeInstallJarFile, forgeVersionJsonFile, "version.json")
	}
	
	val forgeVersionJsonObject: ForgeVersionJsonObject
		get() = forgeVersionJsonFile.readFileToClass(forgeJsonObjectGsonBuilder)
	
	val forgeInstallProfileJsonObject: ForgeInstallProfileJsonObject
		get() = forgeInstallProfileFile.readFileToClass(serializeGameRuntimeDependencyGsonBuilder)
	
	override val modInstallTask: ModInstallTaskInterface
		get() = ForgeInstallTask(version, modVersion)
	override val modKind: KindOfMod
		get() = KindOfMod.FORGE
	
	override fun toString(): String {
		return "ForgeDownloadInfo(modVersion='$modVersion')"
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is ForgeDownloadInfo) return false
		
		if (modVersion != other.modVersion) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		return modVersion.hashCode()
	}
	
	
}
